### PR TITLE
Simplified model for Audience; Leverage join signal's referenceSequenceNumber for catch up logic

### PR DESCRIPTION
**Pull request by vladsud**
Final break out from https://github.com/microsoft/FluidFramework/pull/12042

**Most important changes:**
1. **"self" will show up at the same time in Quorum and Audience** for "write" connection. This is done by ignoring signals for such connections and duplicating Quorum changes into Audience. This gives clients simpler picture to work with.
   - I'd think it makes sense to merge these two data structures in the future, and instead of having 2 of them, provide various adapters / filters that keep usage simple for scenarios where someone needs only "write" clients (as an example).
2. **"connected" transition will happen after "self" shows up in Audience**. This significantly simplifies programming model for users
   - In case of failure (currently defined as 5 seconds of not receiving join signal), we force reconnect. It's hard to say how often it happens and if such recovery is too expensive / has good success rates. If tracking performance of join op any indicator, likely such recovery will not be very successful, but tracking it in telemetry will help us find real bugs in relay service.

Two main pieces **feedback** I'm looking for:
- What to do with NoDeltaConnection
- What to do in rare cases where we do not get join signal at all.
   - I'm planning to use stress test feedback to make a final call here.

**Back-compat considerations:**
- There should be no back-compat concerns here, as behavior (where changed) was never defined. I.e., order in which clients show up in Audience and Quorum are undefined.
- Similar, order of "connected" transition and presence of "self" in Audience is undefined.

Note: depending on results, I may check in this code dark (i.e. flip default feature gate value) and keep working on enabling it.

### CODE REVIEW

File: packages/loader/container-loader/src/connectionManager.ts [hunk 1]
```diff
@@ -81,10 +81,19 @@ function getNackReconnectInfo(nackContent: INackContent) {
  * Implementation of IDocumentDeltaConnection that does not support submitting
  * or receiving ops. Used in storage-only mode.
  */
+const clientNoDeltaStream: IClient = {
+    mode: "read",
+    details: { capabilities: { interactive: true } },
+    permission: [],
+    user: { id: "unknown" }, // we need some "fake" ID here.
```
Comment (vladsud) [inline packages/loader/container-loader/src/connectionManager.ts:88]: @anthony-murphy - curious in your feedback. I want to deliver much simpler programming model, where "connected" means "self" in in audience. But NoDeltaStream breaks that promise. I can do either hack like you see here or exclude specifically NoDeltaStream in connection handler and not wait for join signal. The problem with later - it breaks that simple promise and pushes complexity to consumers. So, I'd prefer the simpler solution (what you see here), but that requires generation of fake user :(

Thoughts?

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 2]
```diff
@@ -333,9 +355,11 @@ class ConnectionStateHandler implements IConnectionStateHandler {
             this.setConnectionState(ConnectionState.Connected);
         } else {
             // Adding this event temporarily so that we can get help debugging if something goes wrong.
+            // We may not see any ops due to being disconnected all that time - that's not an error!
+            const error = source === "timeout" && this.connectionState !== ConnectionState.Disconnected;
```
Comment (vladsud) [inline packages/loader/container-loader/src/connectionStateHandler.ts:359]: FYI - I believe it is (unrelated) bug - we should not report error on timeout if we are disconnected. Even if we are connected, it's a bit moot, as we might have just connected and have not had time to observe leave op/join (i..e it might be general networking issue). But it's hard to make this check that generic, so at least covering most likely case

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 3]
```diff
@@ -356,26 +380,49 @@ class ConnectionStateHandler implements IConnectionStateHandler {
...
         this.setConnectionState(ConnectionState.Disconnected, reason);
     }
 
+    private shouldWaitForJoinSignal() {
+        // No connection - no wait.
+        if (this.connection === undefined) {
+            return false;
+        }
+
+        /* We can go this route, or have NoDeltaStream.initialClients contains "self"
```
Comment (vladsud) [inline packages/loader/container-loader/src/connectionStateHandler.ts:393]: Likely to be removed - see comment in NoDeltaConnection area

**Comment by vladsud** [comment]
Ping - would love to keep moving in making Audience & connectivity transitions better :)

**Comment by markfields** [comment]
Sorry, wasn't able to look today, will set another reminder for Monday!

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 4]
```diff
@@ -242,11 +265,8 @@ class ConnectionStateHandler implements IConnectionStateHandler {
             },
         );
 
-        // Based on recent data, it looks like majority of cases where we get stuck are due to really slow or
-        // timing out ops fetches. So attempt recovery infrequently. Also fetch uses 30 second timeout, so
-        // if retrying fixes the problem, we should not see these events.
         this.joinOpTimer = new Timer(
-            JoinOpTimeoutMs,
+            JoinOpTimeoutMs, // default value is not used - startJoinOpTimer() explicitly provides timeout
```
Comment (markfields) [inline packages/loader/container-loader/src/connectionStateHandler.ts:269]: FYI on the existence of TimerWithNoDefaultTimeout.  Currently lives in garbageCollection.ts but now that there's two spots that could use it, maybe it can move somewhere.  Or you may opt to just user Timer directly without enforcement that start is passed a timeout. In my case I wanted to make sure all callers specified the timeout.

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 5]
```diff
@@ -356,26 +380,49 @@ class ConnectionStateHandler implements IConnectionStateHandler {
...
+            return false;
+        }
+        */
+
+        // Pending clientId could have joined already (i.e. join op/signal already processed).
+        // We are fetching ops from storage in parallel to connecting to Relay Service,
+        // and given async processes, it's possible that we have already processed our own join message before
+        // connection was fully established.
+        // If protocol is not initialized yet, receivedAddMemberEvent() will be called by initProtocol()
+        // later in boot sequence if needed.
```
Comment (markfields) [inline packages/loader/container-loader/src/connectionStateHandler.ts:404]: Nit: Seems irrelevant / out of context details for this next line of code

File: packages/loader/container-loader/src/connectionStateHandler.ts (hunk 4 above)
Comment (markfields) [reply packages/loader/container-loader/src/connectionStateHandler.ts:269]: Also if you leave it as-is I'd pass 0, it'll make the bug clearer if someone doesn't pass a value to start.

File: packages/loader/container-loader/src/connectionStateHandler.ts (hunk 2 above)
Comment (markfields) [reply packages/loader/container-loader/src/connectionStateHandler.ts:359]: Yeah I remember seeing some predictable cases here in Kusto when I was learning this code for the first time. This makes sense.

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 6]
```diff
@@ -264,9 +284,11 @@ class ConnectionStateHandler implements IConnectionStateHandler {
         );
     }
 
-    private startJoinOpTimer() {
+    private startJoinOpTimer(writeConnection: boolean) {
```
Comment (markfields) [inline packages/loader/container-loader/src/connectionStateHandler.ts:287]: I think this param isn't necessary - can use this.connection, right?

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 7]
```diff
@@ -476,18 +520,25 @@ class ConnectionStateHandler implements IConnectionStateHandler {
     // Helper method to switch between quorum and audience.
     // Old design was checking only quorum for "write" clients.
     // Latest change checks audience for all types of connections.
-    protected get membership() {
-        return this.protocol?.quorum;
+    protected get membership(): IMembership | undefined {
+        // It should not matter if we use Audience when this.readClientsWaitForJoinSignal === false.
+        // But only if it's superset of quorum, i.e. when filters to "write" clients, they are always identical!
+        // It's safer to assume that we have bugs and engaging kill-bit switch should bring us back to well-known
+        // and tested state!
```
Comment (markfields) [inline packages/loader/container-loader/src/connectionStateHandler.ts:527]: Took me several passes to realize what this comment was addressing.  Editorial suggestion if you want would be to start like this:

> We could always use audience here, and in practice it will probably be correct.

Then I think what you're saying is that confidence is lower on fidelity of Audience without the readClientsWaitForJoinSignal flag, so it's better to only use Audience when this is enabled. Is that right?

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 8]
```diff
@@ -476,18 +520,25 @@ class ConnectionStateHandler implements IConnectionStateHandler {
...
+        return this.readClientsWaitForJoinSignal ? this.protocol?.audience : this.protocol?.quorum;
     }
 
     public initProtocol(protocol: IProtocolHandler) {
         this.protocol = protocol;
 
-        this.membership?.on("addMember", (clientId) => {
+        this.membership?.on("addMember", (clientId, details) => {
+            assert((details as IClient).mode === "read" || this.protocol?.quorum.getMember(clientId) !== undefined,
```
Comment (markfields) [inline packages/loader/container-loader/src/connectionStateHandler.ts:535]: Maybe do `protocol.quorum` to avoid the `?` which spooked me on first read

File: packages/loader/container-loader/src/connectionManager.ts (hunk 1 above)
Comment (markfields) [reply packages/loader/container-loader/src/connectionManager.ts:88]: I don't mind the fake value, although I haven't scanned through the code to see how user id is used.

Comment (markfields) [reply packages/loader/container-loader/src/connectionManager.ts:88]: I would put "storage-only" or "No Delta Stream" or something in the value though, I guess.

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 9]
```diff
@@ -264,9 +284,11 @@ class ConnectionStateHandler implements IConnectionStateHandler {
         );
     }
 
-    private startJoinOpTimer() {
+    private startJoinOpTimer(writeConnection: boolean) {
         assert(!this.joinOpTimer.hasTimer, 0x234 /* "has joinOpTimer" */);
-        this.joinOpTimer.start();
+        this.joinOpTimer.start(
+            writeConnection ? JoinOpTimeoutMs : JoinSignalTimeoutMs,
```
Comment (markfields) [inline packages/loader/container-loader/src/connectionStateHandler.ts:290]: For a write connection does addMember not fire when the join signal arrives, but only when the join op is processed?

Comment (markfields) [reply packages/loader/container-loader/src/connectionStateHandler.ts:290]: Ok yeah I found the code. Makes sense to stay tied to quorum for write.  Will be great when the future idea of unifying those datastructures is done!

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 10]
```diff
@@ -229,6 +251,7 @@ class ConnectionStateHandler implements IConnectionStateHandler {
 
     constructor(
         private readonly handler: IConnectionStateHandlerInputs,
+        private readonly readClientsWaitForJoinSignal: boolean,
         private _clientId?: string,
```
Comment (markfields) [inline packages/loader/container-loader/src/connectionStateHandler.ts:255]: Unrelated to this change, but I got confused about _clientId.  Maybe name the ctor parameter `clientIdFromPausedSession` or something different, since the parameter has that specific meaning, and in the regular flow it is undefined (but `this._clientId` semantics are the same for regular sessions and deserialized sessions alike)

File: packages/loader/container-loader/src/connectionStateHandler.ts (hunk 7 above)
Comment (markfields) [reply packages/loader/container-loader/src/connectionStateHandler.ts:527]: Could be clearer it's used as a kill switch to actually pull into an if block `if (!flag) { return quorum } return audience`

File: packages/loader/container-loader/src/test/connectionStateHandler.spec.ts [hunk 11]
```diff
@@ -212,19 +298,44 @@ describe("ConnectionStateHandler Tests", () => {
...
             undefined);
 
-        client.mode = "write";
+        connectionDetails.mode = "write";
         assert.strictEqual(connectionStateHandler.connectionState, ConnectionState.Disconnected,
             "Client should be in disconnected state");
-        connectionStateHandler.receivedConnectEvent(client.mode, connectionDetails);
+        connectionStateHandler.receivedConnectEvent(connectionDetails);
         assert.strictEqual(connectionStateHandler.connectionState, ConnectionState.CatchingUp,
             "Client should be in connecting state");
```
Comment (markfields) [inline packages/loader/container-loader/src/test/connectionStateHandler.spec.ts:341]: Next few lines: In practice how could receivedAddMemberEvent be called before initProtocol?

File: packages/loader/container-loader/src/test/connectionStateHandler.spec.ts [hunk 12]
```diff
@@ -374,38 +480,40 @@ describe("ConnectionStateHandler Tests", () => {
...
         assert(connectionStateHandler.pendingClientId !== undefined, "pendingClientId should be set after receiving 'connect' event");
 
         connectionStateHandler.receivedDisconnectEvent("test");
         assert(connectionStateHandler.pendingClientId === undefined, "pendingClientId should not be set after receiving 'disconnect' event");
     });
 
-    it("Should wait for client 1 to leave before moving to connected state(Client 3) when client 2 got disconnected from connecting state", async () => {
-        client.mode = "write";
-        connectionStateHandler.receivedConnectEvent(client.mode, connectionDetails);
+    async function testComplex(mode: ConnectionMode) {
```
Comment (markfields) [inline packages/loader/container-loader/src/test/connectionStateHandler.spec.ts:491]: nit: call it client3Mode

File: packages/loader/container-loader/src/test/connectionStateHandler.spec.ts [hunk 13]
```diff
@@ -146,41 +145,127 @@ describe("ConnectionStateHandler Tests", () => {
         deltaManagerForCatchingUp = new MockDeltaManagerForCatchingUp();
 
         connectionStateHandler = createHandler(
```
Comment (markfields) [inline packages/loader/container-loader/src/test/connectionStateHandler.spec.ts:147]: Observation: The instances created here in `beforeEach` are not disposed.  Actually, none are disposed. So we're leaking a bunch of Timers, right?

And I think before calling dispose you've gotta disconnect or else dispose will assert.

**Review by markfields — APPROVED** [review]
Makes sense!  Please take a look at my comments, but I will sign off now to unblock you.

File: packages/loader/container-loader/src/test/connectionStateHandler.spec.ts (hunk 11 above)
Comment (vladsud) [reply packages/loader/container-loader/src/test/connectionStateHandler.spec.ts:341]: It took me a while to remember, but I think here is what I'm trying to do here:
First of all, this does not result in a call to ConnectionStateHandler, it only sets up quorum state.
There is a tiny tiny race possible, where these events happen in this order:
1.a connection is established (no "cached" mode is used, so it happens in parallel / faster than other steps)
2. some other client produces a summary
3. we get "lucky" and load from that summary as our initial snapshot
4. ConnectionStateHandler.initProtocol is called, "self" is already in the quorum.

We can prevent that case by reordering # 1 & # 3 in Container.load().

Thoughts? Meanwhile, I'll add comments for that sequence here. and in the code

File: packages/loader/container-loader/src/test/connectionStateHandler.spec.ts (hunk 13 above)
Comment (vladsud) [reply packages/loader/container-loader/src/test/connectionStateHandler.spec.ts:147]: I think that does not matter as we never leave it in the state (at the end of test) with timer on (otherwise we would see Node process hanging, I think). But I'll add dispose to protect against future changes

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 14]
```diff
@@ -426,6 +466,10 @@ class ConnectionStateHandler implements IConnectionStateHandler {
 
         const oldState = this._connectionState;
         this._connectionState = value;
+
+        // This is the only place in code that deals with quorum. The rest works with audience
+        // The code below ensures that we do not send ops until we know that old "write" client's disconnect
```
[the remark this answers is not in the document]
Comment (vladsud) [reply packages/loader/container-loader/src/connectionStateHandler.ts:471]: I have some ideas on how to make this logic simpler, faster and not care about old clientId leaving in time.
https://dev.azure.com/fluidframework/internal/_workitems/edit/2217 or some variation of it. We will get there, be patient :)

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 15]
```diff
@@ -297,8 +319,8 @@ class ConnectionStateHandler implements IConnectionStateHandler {
         if (clientId === this.pendingClientId) {
             if (this.joinOpTimer.hasTimer) {
                 this.stopJoinOpTimer();
-            } else {
-                // timer has already fired, meaning it took too long to get join on.
+            } else if (this.shouldWaitForJoinSignal()) {
```
[the remark this answers is not in the document]
Comment (vladsud) [reply packages/loader/container-loader/src/connectionStateHandler.ts:325]: Hm, you are right - good catch! And I think the old telemetry was not right. It's not a lot of garbage in telemetry, but it sounds like we were firing "ReceivedJoinOp" in cases where we did not fire "NoJoinOp", i.e. were we had clientId in quorum before getting connection event - that's only ~50 hits in 30 days (and some of that is likely due to 30-day cutoff in Kusto, i.e. some sessions are truncated), but it's a small bug nonetheless.

union Office_Fluid_FluidRuntime_*
| where Data_eventName contains "JoinOp"
| summarize countif(Data_eventName == "fluid:telemetry:DeltaManager:NoJoinOp"), countif(Data_eventName == "fluid:telemetry:DeltaManager:ReceivedJoinOp") by Data_containerId
| summarize count() by countif_, countif_1
| where countif_1 > countif_

The problem is - with this.readClientsWaitForJoinSignal, I'm making it worse, so need to rethink it.

[the remark this answers is not in the document]
Comment (vladsud) [reply packages/loader/container-loader/src/connectionStateHandler.ts:325]: Refactored code for shouldWaitForJoinSignal() to have slightly different meaning and make this code correct.
